File: PythonOpenGL/step_4/step_4.py
import pygame
from pygame.locals import *
from OpenGL.GL import *
from OpenGL.GLU import *


def draw_line(start, end, color):
    """Draw a line between two points with the given color."""
    glColor3f(*color)
    glBegin(GL_LINES)
    glVertex2f(*start)
    glVertex2f(*end)
    glEnd()


def draw_point(position, color, size=10):
    """Draw a point at the given position with the given color and size."""
    glColor3f(*color)
    glPointSize(size)
    glBegin(GL_POINTS)
    glVertex2f(*position)
    glEnd()


def draw_arrowhead(x, y, isXAxis, size=0.3):
    """Draw a triangle arrowhead for an axis."""
    glColor3f(0, 0, 0)
    glBegin(GL_TRIANGLES)
    if isXAxis:
        glVertex2f(x, y)
        glVertex2f(x - size, y - size / 2)
        glVertex2f(x - size, y + size / 2)
    else:
        glVertex2f(x, y)
        glVertex2f(x - size / 2, y - size)
        glVertex2f(x + size / 2, y - size)
    glEnd()


def setup_opengl():
    """Setup OpenGL for 2D rendering."""
    glClearColor(1.0, 1.0, 1.0, 1.0)  # White background
    glMatrixMode(GL_PROJECTION)
    glLoadIdentity()
    gluOrtho2D(-6, 6, -6, 6)  # Set the coordinate system
    glMatrixMode(GL_MODELVIEW)


def display_everything():
    draw_line((-6, 0), (6, 0), (0, 0, 0))  # X-axis
    draw_arrowhead(6, 0, True)  # Arrowhead for X-axis
    draw_line((0, -6), (0, 6), (0, 0, 0))  # Y-axis
    draw_arrowhead(0, 6, False)  # Arrowhead for Y-axis

    # Draw the red line
    red_start = (-6.0, 2.0 / 3.0)
    red_end = (6.0, 26.0 / 3.0)
    draw_line(red_start, red_end, (1.0, 0.0, 0.0))

    # Draw the green line
    green_start = (-6.0, 4.0)
    green_end = (6.0, 4.0)
    draw_line(green_start, green_end, (0.0, 1.0, 0.0))

    # Draw the blue point
    draw_point((-1.0, 4.0), (0.0, 0.0, 1.0))

def main():
    # Initialize pygame and set up an OpenGL context
    pygame.init()
    display = (800, 600)
    pygame.display.set_mode(display, DOUBLEBUF | OPENGL)
    pygame.display.set_caption("2D Line Plot")

    setup_opengl()

    running = True
    while running:
        for event in pygame.event.get():
            if event.type == QUIT:
                running = False

        # Clear the screen
        glClear(GL_COLOR_BUFFER_BIT)

        display_everything()

        pygame.display.flip()
        pygame.time.wait(10)

    pygame.quit()



main()

File: Gauss_ogl_Begis_Maulenbergenov/step_4/step4.cpp
#include <SDL/SDL.h>
#include <GL/gl.h>
#include <GL/glu.h>

// Function to draw lines
void Draw_line(float V1[3], float V2[3]) {
    glBegin(GL_LINES);
    glVertex3f(V1[0], V1[1], V1[2]);
    glVertex3f(V2[0], V2[1], V2[2]);
    glEnd();
}

// Function to draw triangle arrowheads
void Draw_triangle(float x, float y, bool isXAxis) {
    glBegin(GL_TRIANGLES);
    if (isXAxis) {
        glVertex3f(x, y, 0.0f);
        glVertex3f(x - 0.6f, y - 0.4f, 0.0f);
        glVertex3f(x - 0.6f, y + 0.4f, 0.0f);
    } else { // Y axis
        glVertex3f(x, y, 0.0f);
        glVertex3f(x - 0.3f, y - 0.6f, 0.0f);
        glVertex3f(x + 0.3f, y - 0.6f, 0.0f);
    }
    glEnd();
}

// Initialize OpenGL
void init() {
    glClearColor(1.0, 1.0, 1.0, 1.0);
    glMatrixMode(GL_PROJECTION);
    glLoadIdentity();
    gluOrtho2D(-8.0, 8.0, -8.0, 8.0);
    glMatrixMode(GL_MODELVIEW);
}

// Render axes and arrowheads
void display() {
    glClear(GL_COLOR_BUFFER_BIT);

    glLoadIdentity();

    // Draw X-axis
    glLineWidth(2.0);
    glColor3f(0.0, 0.0, 0.0); // Black
    float V1[3] = {-8.0f, 0.0f, 0.0f};
    float V2[3] = {8.0f, 0.0f, 0.0f};
    Draw_line(V1, V2);

    // Arrowhead for X-axis
    Draw_triangle(8.0f, 0.0f, true);

    // Draw Y-axis
    V1[0] = 0.0f;
    V1[1] = -8.0f;
    V2[0] = 0.0f;
    V2[1] = 8.0f;
    Draw_line(V1, V2);

    // Arrowhead for Y-axis
    Draw_triangle(0.0f, 8.0f, false);

    glLineWidth(5.0);
    glColor3f(1.0, 0.0, 0.0);
    float redStart[3] = {-8.0f, -7.0f / 3.0f, 0.0f};
    float redEnd[3] = {8.0f, 25.0f / 3.0f, 0.0f};
    Draw_line(redStart, redEnd);


    glColor3f(0.0, 1.0, 0.0);
    float greenStart[3] = {-8.0f, 5.0f, 0.0f};
    float greenEnd[3] = {8.0f, 5.0f, 0.0f};
    Draw_line(greenStart, greenEnd);


    glEnable(GL_POINT_SMOOTH);
    glHint(GL_POINT_SMOOTH_HINT, GL_NICEST);
    glColor3f(0.0, 0.0, 1.0);
    glPointSize(12.0);
    glBegin(GL_POINTS);
    glVertex3f(3.0f, 5.0f, 0.0f);
    glEnd();
    glDisable(GL_POINT_SMOOTH);

    SDL_GL_SwapBuffers();
}

int main(int argc, char* args[]) {
    SDL_Init(SDL_INIT_EVERYTHING);
    SDL_SetVideoMode(300, 200, 32, SDL_SWSURFACE | SDL_OPENGL);

    int running = 1;
    SDL_Event event;

    init();

    while (running) {
        while (SDL_PollEvent(&event)) {
            if (event.type == SDL_QUIT) {
                running = 0;
            }
        }
        display();
    }

    SDL_Quit();
    return 0;
}
